### PR TITLE
Fix memory corruption issue in emmalloc implementation of memalign() 

### LINKED ISSUE #10363: emmalloc assertion error: region_is_in_use
Using the latest upstream of emsdk, I got this exception. It seems like a bug in the recent update of emmalloc? sbrk is also used by bdwgc.

image

### PR DESCRIPTION
**Pull request by juj**
Fix memory corruption issue in emmalloc implementation of memalign(). Closes #10363.

### CODE REVIEW

File: system/lib/emmalloc.cpp [hunk 1]
```diff
@@ -354,10 +354,23 @@ static void resize_region(void *ptr, uint32_t size)
...
+  // Free regions have all the size bits reversed (highest bit tells it is a free region)
+  uint32_t usedMask = (uint32_t)(((int32_t)oldSize) >> 31); // Sign extend the highest bit to whole word
+
+  // Write new region size:
   *(uint32_t*)ptr = size;
-  uint32_t *sizeAtEnd = (uint32_t*)ptr + (size>>2) - 1;
-  uint32_t usedMask = (*(int32_t*)sizeAtEnd) >> 31;
-  *sizeAtEnd = size ^ usedMask;
+  uint32_t *newRegionEnd = (uint32_t*)((uintptr_t)ptr + size);
+  newRegionEnd[-1] = size ^ (usedMask >> 31);
```
Comment (kripken) [inline system/lib/emmalloc.cpp]: `usedMask` is either all 0s or all 1s, I believe? If so then `usedMask >> 31` is either 0 or 1. Is that right? That seems to affect only the lowest bit in the XOR here.

Comment (juj) [reply system/lib/emmalloc.cpp]: Err, you are absolutely right.. had an excess >> 31 there.

However now that I look at this with fresh eyes in the morning and wonder why this ever worked, I realize that this function was originally used to resize both free and used regions, but after a change at some point, this function is now only ever called to resize used regions. That is why everything passes with this PR: `usedMask` is always all zeros.

Hence we can simplify and just nuke the whole function..

File: system/lib/emmalloc.cpp [hunk 2]
```diff
@@ -905,6 +896,16 @@ void emmalloc_free(void *ptr)
   MALLOC_ACQUIRE();
 
   uint32_t size = region->size;
+#ifdef EMMALLOC_DEBUG_LOG
+  if (size < sizeof(Region) || !region_is_in_use(region))
+  {
+    if (debug_region_is_consistent(region))
+      // LLVM wasm backend bug: cannot use MAIN_THREAD_ASYNC_EM_ASM() here, that generates internal compiler error
```
Comment (kripken) [inline system/lib/emmalloc.cpp]: I believe this bug can be worked around with having a single EM_ASM and passing into it a string for the initial differing text. As the rest is the same, that would also be shorter here in the code.

Comment (juj) [reply system/lib/emmalloc.cpp]: If I pass it as a string, I need to convert it from Wasm string to JS string, which is kind of dirty... I think I'll leave it like that.

**Pull request merged**